### ISSUE TITLE
The code proxier.go failed to check  kernel module when the folder  /proc/module is not exsit

### ISSUE THREAD

**Issue by robinchia** [kubernetes/kubernetes#108579]
### What happened?

When the kernel is statically compiled and statically loaded, and the folder /proc/modules is not exsit. In the kubernetes, use the ipvs. The code proxier.go failed to check  /proc/module  when it is not exsit. In the line 628, the err is not nil. But when  the folder /proc/modules is exsit, it go to the line 639. 
The question is that, how the code runs to 614 line?

### What did you expect to happen?

In the kubernetes, use the ipvs. the kernel is statically compiled and statically loaded. It check out the config in the folder /boot, such as /boot/config-****

### How can we reproduce it (as minimally and precisely as possible)?

If you choose an arm64 machine,  config the kernel config , set CONFIG_MODULES=no, compile statically the kernel. Join the node into k8s's cluster. So you will see this in the pod kube-proxy***.

### Kubernetes version

<details>

```console
$ kubectl version
Client Version: version.Info{Major:"1", Minor:"19", GitVersion:"v1.19.16", GitCommit:"e37e4ab4cc8dcda84f1344dda47a97bb1927d074", GitTreeState:"clean", BuildDate:"2021-10-27T16:25:59Z", GoVersion:"go1.15.15", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"19", GitVersion:"v1.19.16", GitCommit:"e37e4ab4cc8dcda84f1344dda47a97bb1927d074", GitTreeState:"clean", BuildDate:"2021-10-27T16:20:18Z", GoVersion:"go1.15.15", Compiler:"gc", Platform:"linux/amd64"}
```

</details>

### Cloud provider

<details>
There is not a cloud provider. This is for myself.
</details>

### OS version

<details>

```console
# On Linux:
$ cat /etc/os-release
PRETTY_NAME="Debian GNU/Linux 10 (buster)"
NAME="Debian GNU/Linux"
VERSION_ID="10"
VERSION="10 (buster)"
VERSION_CODENAME=buster
ID=debian
HOME_URL="https://www.debian.org/"
SUPPORT_URL="https://www.debian.org/support"
BUG_REPORT_URL="https://bugs.debian.org/"
$ uname -a
Linux armk1101 4.9.216-q4_2107u2

```

</details>

### Install tools

<details>

</details>

### Container runtime (CRI) and version (if applicable)

<details>

</details>

### Related plugins (CNI, CSI, ...) and versions (if applicable)

<details>

</details>

**Comment by mk46** [comment]
Hi @aojea, if https://github.com/mk46/kubernetes/commit/d3bcf87da345a3714782305747ad5ba65de3d82a make sense then I would like to open a PR for this.
 Thanks!!

**Comment by aojea** [comment]
> The code proxier.go failed to check /proc/module when it is not exsit. In the line 628, the err is not nil. But when the folder /proc/modules is exsit, it go to the line 639.
> The question is that, how the code runs to 614 line?

Have we figured out the answer to the question?
Do we have an unit test reproducing the issue?

https://github.com/kubernetes/kubernetes/blob/19762408612e49401aa46d8396c7032462fe5ab0/pkg/proxy/ipvs/proxier.go#L618-L637

/area ipvs

**Comment by robinchia** [comment]
I think the question is that,  when I use statically compiled  kernel system as cabernets work node, and the folder /proc/module is not exsit. The Kube-proxy log is that:
E0308 05:15:01.994259       1 server_others.go:106] Can't use the IPVS proxier: error getting installed ipvs required kernel modules: Failed to read file /proc/modules with error open /proc/modules: no such file or directory
I0308 05:15:02.030448       1 node.go:136] Successfully retrieved node IP: 10.2.11.1
I0308 05:15:02.030553       1 server_others.go:142] kube-proxy node IP is an IPv4 address (10.2.11.1), assume IPv4 operation
I0308 05:15:02.086917       1 server_others.go:185] Using iptables Proxier.
I0308 05:15:02.087924       1 server.go:650] Version: v1.19.16
I0308 05:15:02.089195       1 conntrack.go:52] Setting nf_conntrack_max to 262144
I0308 05:15:02.089888       1 config.go:315] Starting service config controller
I0308 05:15:02.089948       1 shared_informer.go:240] Waiting for caches to sync for service config
I0308 05:15:02.090117       1 config.go:224] Starting endpoint slice config controller
I0308 05:15:02.090167       1 shared_informer.go:240] Waiting for caches to sync for endpoint slice config
I0308 05:15:02.190384       1 shared_informer.go:247] Caches are synced for service config
I0308 05:15:02.190423       1 shared_informer.go:247] Caches are synced for endpoint slice config
And when I use statically compiled  kernel system as cabernets work node, the folder /proc/module is exsit. The log is this:
W0308 06:59:32.826787       1 proxier.go:639] Failed to read file /lib/modules/4.9.216-q4_2107u2/modules.builtin with error open /lib/modules/4.9.216-q4_2107u2/modules.builtin: no such file or directory. You can ignore this message when kube-proxy is running inside container without mounting /lib/modules
W0308 06:59:32.833873       1 proxier.go:649] Failed to load kernel module ip_vs with modprobe. You can ignore this message when kube-proxy is running inside container without mounting /lib/modules
W0308 06:59:32.836506       1 proxier.go:649] Failed to load kernel module ip_vs_rr with modprobe. You can ignore this message when kube-proxy is running inside container without mounting /lib/modules
W0308 06:59:32.839137       1 proxier.go:649] Failed to load kernel module ip_vs_wrr with modprobe. You can ignore this message when kube-proxy is running inside container without mounting /lib/modules
W0308 06:59:32.841857       1 proxier.go:649] Failed to load kernel module ip_vs_sh with modprobe. You can ignore this message when kube-proxy is running inside container without mounting /lib/modules
W0308 06:59:32.844632       1 proxier.go:649] Failed to load kernel module nf_conntrack_ipv4 with modprobe. You can ignore this message when kube-proxy is running inside container without mounting /lib/modules
E0308 06:59:32.844833       1 server_others.go:106] Can't use the IPVS proxier: IPVS proxier will not be used because the following required kernel modules are not loaded: [ip_vs ip_vs_rr ip_vs_wrr ip_vs_sh]
I0308 06:59:32.877082       1 node.go:136] Successfully retrieved node IP: 10.2.11.1
I0308 06:59:32.877153       1 server_others.go:142] kube-proxy node IP is an IPv4 address (10.2.11.1), assume IPv4 operation
I0308 06:59:32.928535       1 server_others.go:185] Using iptables Proxier.
I0308 06:59:32.928959       1 server.go:650] Version: v1.19.16
I0308 06:59:32.930309       1 conntrack.go:52] Setting nf_conntrack_max to 262144
I0308 06:59:32.930649       1 config.go:315] Starting service config controller
I0308 06:59:32.930662       1 shared_informer.go:240] Waiting for caches to sync for service config
I0308 06:59:32.930699       1 config.go:224] Starting endpoint slice config controller
I0308 06:59:32.930711       1 shared_informer.go:240] Waiting for caches to sync for endpoint slice config
I0308 06:59:33.031438       1 shared_informer.go:247] Caches are synced for service config
I0308 06:59:33.031799       1 shared_informer.go:247] Caches are synced for endpoint slice config

I think the code question is in 620 line. The sentence "err == os.ErrNotExist"  should judge that is true when the folder /proc/module is exsit, the code should go to code 621 line and so on. Am I right ？

**Comment by uablrek** [comment]
IMHO making assumptions on how the kernel is built is wrong. K8s should test the _function_ not check for modules or /proc/config. The problem is general and not specific for kube-proxy. There _may_ be cases when checking the modules is the only option, but then there should be an option to disable the tests and leave the responsibility to the user, e.g `--disable-kernel-checks`. Users that have this problem build their own kernel and should know what they are doing.

I used a statically linked kernel but has been forced to use modules. Since I want to be able to switch kernel to kvm without replacing the disk (containing the the matching modules) I store the modules in a "initfs" that is included (as a cpio) in the `bzimage`.

**Comment by robinchia** [comment]
> > When the kernel is statically compiled and statically loaded, and the folder /proc/modules is not exsit. In the kubernetes, use the ipvs. The code proxier.go failed to check /proc/module when it is not exsit. In the line 628, the err is not nil. But when the folder /proc/modules is exsit, it go to the line 639. The question is that, how the code runs to 614 line?
> 
> Line 614 always contains a string array of
> 
> https://github.com/kubernetes/kubernetes/blob/19762408612e49401aa46d8396c7032462fe5ab0/pkg/util/ipvs/ipvs.go#L74-L88
> 
> 
> by calling
> https://github.com/kubernetes/kubernetes/blob/19762408612e49401aa46d8396c7032462fe5ab0/pkg/util/ipvs/ipvs.go#L126-L134
> 
> I don't think line 628 will cause any regex parsing error while parsing the above kernel module. @robinchia Kindly share the log for the same.

There is not any regex parsing error in line 628. In my case, n matter whether the /proc/modules directory exists or not, this code will not detect the configuration file which is  /boot/config-XXX, that makes it impossible to read the loaded features of the static kernel.
The code won't run to there:
`kernelConfigFile := fmt.Sprintf("/boot/config-%s", kernelVersionStr) 
  kConfig, err := ioutil.ReadFile(kernelConfigFile) 
...
`

**Comment by robinchia** [comment]
> > And when I use statically compiled kernel system as cabernets work node, the folder /proc/module is exsit. The log is this:
> 
> why should that exist on the statically compiled kernel? (is a question, I don't know if that is valid) I think that is the problem, the code assume that if the `/proc/module` folder exists then the modules must exist

I'm sorry that it is /proc/modules, not /proc/module.
If I compile kernel when the kernel's config is CONFIG_MODULES = Y, so the /proc/modules will not exist.

**Comment by dcbw** [comment]
@robinchia is the problem that if you build with CONFIG_MODULES=y, but CONFIG_IPVS_*=y that /proc/modules will exist and the regex will not be used to detect presence of built-in  modules?

**Comment by thockin** [comment]
> IMHO making assumptions on how the kernel is built is wrong. K8s should test the function not check for modules or /proc/config.gz or /boot/config*. 

Exactly right - I am not sure the specifics of this case, but I agree with the principle.

**Comment by robinchia** [comment]
> @robinchia is the problem that if you build with CONFIG_MODULES=y, but CONFIG_IPVS_*=y that /proc/modules will exist and the regex will not be used to detect presence of built-in modules?

@dcbw Partially right. I think that  the value of  "err == os.ErrNotExist" may not be true when the /proc/modules  exist.

**Comment by thockin** [comment]
What's the next step on this?  Should we accept the bug as real?

**Comment by dcbw** [comment]
@robinchia would you be willing to do a PR to work around the issue?

**Comment by robinchia** [comment]
> @robinchia would you be willing to do a PR to work around the issue?

Yes, that is ok. I need some time to Lookup the codes about the issue. Should you tell me that which branch of the kubernetes I need to work around about this?

**Comment by robinchia** [comment]
@dcbw Hi, I fix the bug for the code "kubernetes/pkg/proxy/ipvs/proxier.go" in the branch release-1.19. Maybe it is not in the release-note-label-needed, so I remove it. Which branch should  I fix the bug in ?

**Comment by robinchia** [comment]
@dcbw Can you have time to give me a reply？

**Comment by uablrek** [comment]
Please read:

* https://github.com/kubernetes/community/blob/master/contributors/guide/README.md#contributor-guide

I found this useful;

* https://github.com/kubernetes/community/blob/master/contributors/guide/github-workflow.md

**Comment by robinchia** [comment]
@uablrek Thanks a lot.

**Comment by uablrek** [comment]
#### TL;DR

I propose;

* Ditch the module checks entirely
* Check the `ipvs` *function* by doing a list command (will automativally load the module if needed)
* Assume that the `rr` (round-robin) scheduler exist if ipvs is supported (no explicit check)
* Check the `conntrack` *function* by doing a list command (will automativally load the module if needed)

@thockin @andrewsykim @aojea Please comment on the above.

#### Background

Now we have 2 PR's that adds ways to check modules; https://github.com/kubernetes/kubernetes/pull/110017 and https://github.com/kubernetes/kubernetes/pull/109937.

As mentioned in https://github.com/kubernetes/kubernetes/pull/109937#issuecomment-1123218857 I would like to remove the explicit module check entirely, rather than put another band-aid on a flawed function.

To check the **function** a good way is to make a non-intrusive ipvs operation (not calling any external program), for instance;

https://github.com/kubernetes/kubernetes/blob/b74d023e70d6064c7f3f77031e7d26ec38497fc9/pkg/util/ipvs/ipvs.go#L39

If that succeeds, the ipvs function can be used (proxy-mode=ipvs can be used).

The current function also checks the `rr`, `wrr` and `sh` ipvs schedulers. Only `rr` is used though, other schedulers can be configured explicitly, but IMO they are not needed for the function and must not be checked. IMHO it is safe to assume that if ipvs is in the kernel (statically or by module) then the `rr` sheduler also exist.

The conntracker *module* is also checked. Again the conntrack *function* can be checked with some non-intrusive command, like a list command.

**Comment by robinchia** [comment]
@uablrek 
IMHO it is not a good idea to ditch the module checks entirely. It will need a lot of working for checking the ipvs, rr, wrr, and sh by commands, and so on.

**Comment by danwinship** [comment]
> Now we have 2 PR's that adds ways to check modules; #110017 and #109937.

Three actually:  #97070

I agree that these checks are mostly pointless, and support your proposal.

(Note that running `ipvsadm` and `conntrack` list commands also verifies that the command-line tools are installed, which is something that I think has been a problem more often in real-life than kernel modules.)
